### PR TITLE
Handle multiple awaits and suspend-on-exit for async let tasks.

### PR DESCRIPTION
**Pull request by jckarter**
Change the code generation patterns for `async let` bindings to use an ABI based on the following
functions:

- `swift_asyncLet_start`, which as before, starts an `async let` child task, but which additionally
  now associates the `async let` with a caller-owned buffer to receive the result of the task.
  This is intended to allow the task to emplace its result in caller-owned memory, allowing the
  child task to be deallocated after completion without invalidating the result buffer.
- `swift_asyncLet_get[_throwing]`, which replaces `swift_asyncLet_wait[_throwing]`. Instead of
  returning a copy of the value, this entry point concerns itself with populating the local buffer.
  If the buffer hasn't been populated, then it awaits completion of the task and emplaces the
  result in the buffer; otherwise, it simply returns. The caller can then read the result out of
  its owned memory. These entry points are intended to be used before every read from the
  `async let` binding, after which point the local buffer is guaranteed to contain an initialized
  value.
- `swift_asyncLet_finish`, which replaces `swift_asyncLet_end`. Unlike `_end`, this variant
  is async and will suspend the parent task after cancelling the child to ensure it finishes
  before cleaning up. The local buffer will also be deinitialized if necessary. This is intended
  to be used on exit from an `async let` scope, to handle cleaning up the local buffer if necessary
  as well as cancelling, awaiting, and deallocating the child task.

To avoid an epoch break with existing swiftinterfaces and ABI clients, the old builtins and entry
points are kept intact for now, but SILGen now only generates code using the new interface.

This new interface fixes several issues with the old async let codegen, including use-after-free
crashes if the `async let` was never awaited, and the inability to read from an `async let` variable
more than once.

rdar://77855176

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

File: stdlib/public/Concurrency/AsyncLet.cpp [hunk 1]
```diff
@@ -140,7 +188,84 @@ static void swift_asyncLet_wait_throwingImpl(
...
+  }
+
+  // Mark the async let as having its result populated.
+  // The only task that can ask this of the async let is the same parent task
+  // that's currently executing, so we can set it now and tail-call future_wait,
+  // since by the time we can call back it will be populated.
+  asImpl(alet)->setHasResultInBuffer();
+  swift_task_future_wait(reinterpret_cast<OpaqueValue*>(resultBuffer),
+                         callerContext, alet->getTask(),
+                         resumeFunction, callContext);
```
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncLet.cpp:215]: Right, that should be safe to do 👍

**Review by ktoso — APPROVED** [review]
Read through in depth and looks fantastic as far as I can tell :-)

File: include/swift/ABI/MetadataValues.h [hunk 2]
```diff
@@ -53,8 +53,8 @@ enum {
   /// The number of words in a task group.
   NumWords_TaskGroup = 32,
 
-  /// The number of words in an AsyncLet (flags + task pointer)
-  NumWords_AsyncLet = 8, // TODO: not sure how much is enough, these likely could be pretty small
+  /// The number of words in an AsyncLet (flags + child task context & allocation)
+  NumWords_AsyncLet = 80, // 640 bytes ought to be enough for anyone
```
[the remark this answers is not in the document]
Comment (jckarter) [reply include/swift/ABI/MetadataValues.h:57]: We use the same block for the child task's initial slab, and the previous code gave the child 512 bytes of slab to begin with. 640 seemed like a nice round number to allow for some growth in the task context header overhead and still ensure at least 512 bytes still remain for the child context to start with.

File: stdlib/public/Concurrency/AsyncLet.cpp [hunk 3]
```diff
@@ -119,6 +147,26 @@ void swift::swift_asyncLet_start(AsyncLet *alet,
...
 
+SWIFT_CC(swift)
+void swift::swift_asyncLet_begin(AsyncLet *alet,
+                                 TaskOptionRecord *options,
+                                 const Metadata *futureResultType,
+                                 void *closureEntryPoint,
+                                 HeapObject *closureContext,
+                                 void *resultBuffer) {
+  auto flags = TaskCreateFlags();
+  flags.setEnqueueJob(true);
```
[the remark this answers is not in the document]
Comment (jckarter) [reply stdlib/public/Concurrency/AsyncLet.cpp:158]: Hm, I followed the code that asyncLet_start already had for getting an async let task started. Is that not implied by being an async let task?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test OS X

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by drexin** [comment]
@swift-ci clean test linux

**Comment by drexin** [comment]
@swift-ci Please test Ubuntu 20.04 platform

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by drexin** [comment]
@swift-ci test linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please clean test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please clean test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Comment by aschwaighofer** [comment]
Please test with following PR: 
https://github.com/apple/llvm-project/pull/3121

@swift-ci test linux

**Comment by aschwaighofer** [comment]
Please test with following PR:
apple/llvm-project#3121

@swift-ci test macOS

**Comment by jckarter** [comment]
Please test with following PR:
apple/llvm-project#3121

@swift-ci Please test

**Comment by jckarter** [comment]
Please test with following PR:
apple/llvm-project#3121

@swift-ci Please test

**Comment by jckarter** [comment]
Please test with following PR:
apple/llvm-project#3121

@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test OS X

**Comment by jckarter** [comment]
Testing if it tests cleanly with all of the other memory bugs fixed, but without the LLVM patch:

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Pull request merged**